evalml/automl/automl_search.py

```diff
@@ -740,62 +626,40 @@ def _add_result(self, trained_pipeline, parameters, training_time, cv_data, cv_s
         if self.add_result_callback:
             self.add_result_callback(self._results['pipeline_results'][pipeline_id], trained_pipeline, self)
 
-    def _evaluate_pipelines(self, current_pipeline_batch, X, y, baseline=False, search_iteration_plot=None):
+    def _evaluate_pipelines(self, current_pipeline_batch, X, y, baseline=False, engine=None):
         current_batch_pipeline_scores = []
-        add_single_pipeline = False
-        if isinstance(current_pipeline_batch, PipelineBase):
-            current_pipeline_batch = [current_pipeline_batch]
-            add_single_pipeline = True
-
-        while len(current_pipeline_batch) > 0 and (add_single_pipeline or baseline or self._check_stopping_condition(self._start)):
-            pipeline = current_pipeline_batch.pop()
-            try:
-                parameters = pipeline.parameters
-                logger.debug('Evaluating pipeline {}'.format(pipeline.name))
-                logger.debug('Pipeline parameters: {}'.format(parameters))
-
-                if self.start_iteration_callback:
-                    self.start_iteration_callback(pipeline.__class__, parameters, self)
-                desc = f"{pipeline.name}"
-                if len(desc) > self._MAX_NAME_LEN:
-                    desc = desc[:self._MAX_NAME_LEN - 3] + "..."
-                desc = desc.ljust(self._MAX_NAME_LEN)
-
-                if not add_single_pipeline:
-                    update_pipeline(logger, desc, len(self._results['pipeline_results']) + 1, self.max_iterations,
-                                    self._start, 1 if baseline else self._automl_algorithm.batch_number, self.show_batch_output)
-
-                evaluation_results = self._compute_cv_scores(pipeline, X, y)
-                parameters = pipeline.parameters
-
+        current_pipeline_batch_size = 1 if isinstance(current_pipeline_batch, PipelineBase) else len(current_pipeline_batch)
+        if engine is None:
```

Comment:
@freddyaboulton Re creating a new engine for each batch: it's definitely not the intended behavior. I moved this check into `search` and `add_to_rankings`. I also added engines as an optional parameter in `add_to_rankings`. I think it makes sense to still make the engines parameter optional and keep this check in `_evaluate_pipelines`. In the future, it would make setting `_evaluate_pipelines` as public easier.

@jeremyliweishih Re changing the default parameter: I originally had this check set here rather than as a default parameter so that in subsequent PRs we could refactor this later on to use a Dask engine if supported and move this as a fallback engine. This however is a separate refactor down the line and I wouldn't be opposed to changing the default parameter. 

In terms of adding engine as a default parameter arg, I originally wanted to do this but decided against it.
- Currently, this implementation uses the entire AutoMLSearch object in the engine to process pipelines. If the Dask Engine was part of the automl object, then Dask would not be able to serialize it since it can't serialize itself. This could be resolved by having a preprocessing step in engines itself that would remove the engines attribute, but for implementing additional engines in the future it's pretty unintuitive. 
- While a bit niche, I could see cases where we would want to use different engines for evaluating different types of pipelines. For example, for a large search I might want to use the Dask engine while if I'm adding a single pipeline with `add_to_rankings` I might use the SequentialEngine. 

I do see how passing in the engine could become tedious (especially for subsequent functions that evaluates pipelines such as `add_to_rankings`). 

Again, looking forward to discussing each one of these points further!

Reply:
@christopherbunn Agreed that keeping engine as an optional parameter to `_evaluate_pipelines` is a good idea!